### PR TITLE
Add generators section in kubectl conventions

### PR DESCRIPTION
**Pull request by soltysh**
@janetkuo as proposed in https://github.com/kubernetes/kubernetes/pull/32671#issuecomment-249021014 I'm adding a generator section, ptal.

@devin-donnelly it would make a good candidate for 1.4 release, not sure how far you're with the release?

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/1292)

### CODE REVIEW

File: docs/user-guide/kubectl-conventions.md [hunk 1]
```diff
@@ -8,15 +8,45 @@ assignees:
...
 * Don't rely on context, preferences, or other implicit state
 
+## Generators
+
+`kubectl run` provides following generators:
+
+* `run/v1` - creates a replication controller.
+* `run-pod/v1` - creates a pod.
+* `deployment/v1beta1` - creates a deployment.
+* `job/v1beta1` - creates a job (using `extensions/v1beta1' endpoint).
```
Comment (janetkuo) [inline docs/user-guide/kubectl-conventions.md]: `extensions/v1beta1` (the last symbol isn't correct)

File: docs/user-guide/kubectl-conventions.md [hunk 2]
```diff
@@ -8,15 +8,45 @@ assignees:
...
 
 If you need stable output in a script, you should:
 
-* Request one of the machine-oriented output forms, such as `-o name`, `-o json`, `-o yaml`, `-o go-template`, or `-o jsonpath` 
+* Request one of the machine-oriented output forms, such as `-o name`, `-o json`, `-o yaml`, `-o go-template`, or `-o jsonpath`
 * Specify `--output-version`, since those output forms (other than `-o name`) output the resource using a particular API version
 * Specify `--generator` to pin to a specific behavior forever, if using generator-based commands (such as `kubectl run` or `kubectl expose`)
 * Don't rely on context, preferences, or other implicit state
 
+## Generators
```
Comment (janetkuo) [inline docs/user-guide/kubectl-conventions.md]: Suggest nesting it under Best Practices > kubectl run > Generators (or `kubectl run --generators`, or both)

File: docs/user-guide/kubectl-conventions.md [hunk 3]
```diff
@@ -8,15 +8,45 @@ assignees:
...
+* `run/v1` - creates a replication controller.
+* `run-pod/v1` - creates a pod.
+* `deployment/v1beta1` - creates a deployment.
+* `job/v1beta1` - creates a job (using `extensions/v1beta1' endpoint).
+* `job/v1` - creates a job.
+* `scheduledjob/v2alpha1` - creates a scheduled job.
+
+Following rules are implemented if no generator is specified for cluster version at least 1.3:
+
+* For `--restart=Always' a deployment generator (`deployment/v1beta1`) is chosen.
```
Comment (janetkuo) [inline docs/user-guide/kubectl-conventions.md]: Use ` after Always

File: docs/user-guide/kubectl-conventions.md [hunk 4]
```diff
@@ -8,15 +8,45 @@ assignees:
...
-* Request one of the machine-oriented output forms, such as `-o name`, `-o json`, `-o yaml`, `-o go-template`, or `-o jsonpath` 
+* Request one of the machine-oriented output forms, such as `-o name`, `-o json`, `-o yaml`, `-o go-template`, or `-o jsonpath`
 * Specify `--output-version`, since those output forms (other than `-o name`) output the resource using a particular API version
 * Specify `--generator` to pin to a specific behavior forever, if using generator-based commands (such as `kubectl run` or `kubectl expose`)
 * Don't rely on context, preferences, or other implicit state
 
+## Generators
+
+`kubectl run` provides following generators:
+
```
Comment (janetkuo) [inline docs/user-guide/kubectl-conventions.md]: We should explain what generator is and mention `--generator` flag

File: docs/user-guide/kubectl-conventions.md (hunk 1 above)
Comment (janetkuo) [inline docs/user-guide/kubectl-conventions.md]: Perhaps we should put the object first, generator later, something like:

Use `--generator` flag to specify the object to be generated, for example, to generate a:
* Replication Controller: `--generator=run/v1` (or just `run/v1` is enough?)
* Pod: `--generator=run-pod/v1` 
...

Or maybe use a table?

File: docs/user-guide/kubectl-conventions.md [hunk 5]
```diff
@@ -8,15 +8,45 @@ assignees:
...
+* `scheduledjob/v2alpha1` - creates a scheduled job.
+
+Following rules are implemented if no generator is specified for cluster version at least 1.3:
+
+* For `--restart=Always' a deployment generator (`deployment/v1beta1`) is chosen.
+* For `--restart=OnFailure` a job generator (`job/v1`) is chosen.
+* For `--restart=Never` a pod generator (`run-pod/v1`) is chosen.
+
+Additionally, for cluster version at least 1.4 if `--schedule` flag is set a scheduled job
+generator ('scheduledjob/v2alpha1') is chosen.
```
Comment (janetkuo) [inline docs/user-guide/kubectl-conventions.md]: ` instead of ', same as below

File: docs/user-guide/kubectl-conventions.md [hunk 6]
```diff
@@ -8,15 +8,45 @@ assignees:
...
+
+For cluster version 1.2:
+
+* For `--restart=Always` a deployment generator ('deployment/v1beta1') is chosen.
+* For other restart policy a job generator ('job/v1') is chosen.
+
+For cluster version 1.1 and older:
+
+* For `--restart=Always` a replication controller ('run/v1') is chosen
+* For other restart policy a pod generator ('run-pod/v1') is chosen.
```
Comment (janetkuo) [inline docs/user-guide/kubectl-conventions.md]: Perhaps something like this is easier to read?

|   Generated Resource   | Cluster 1.4         | Cluster 1.3                 | Cluster 1.2                                 | Cluster <=1.1                               |
|:----------------------:|---------------------|---------------------|-------------------------------------|-------------------------------------|
| Scheduled Job          | `--schedule=<cron>`          | N/A                 | N/A                                 | N/A                                 |
| Deployment             | `--restart=Always`    | `--restart=Always`    | `--restart=Always`                    | N/A                                 |
| Replication Controller | `--generator=run/v1`  | `--generator=run/v1`  | `--generator=run/v1`                  | `--restart=Always`                    |
| Job                    | `--restart=OnFailure` | `--restart=OnFailure` | `--restart=OnFailure` OR `--restart=Never` | N/A                                 |
| Pod                    | `--restart=Never`     | `--restart=Never`     | `--generator=run-pod/v1`              | `--restart=OnFailure` OR `--restart=Never` |

This isn't clean enough though, we should also mention that `--schedule` sets the `--generator` for you, as well as `--restart`, and if you set the generator to something else it wouldn't work as expected. Also, once `--schedule` set the `--generator`, `--restart` won't set it again (order: `--schedule` --> `--restart` --> `--generator`).

File: docs/user-guide/kubectl-conventions.md [hunk 7]
```diff
@@ -8,15 +8,45 @@ assignees:
...
 
-* Request one of the machine-oriented output forms, such as `-o name`, `-o json`, `-o yaml`, `-o go-template`, or `-o jsonpath` 
+* Request one of the machine-oriented output forms, such as `-o name`, `-o json`, `-o yaml`, `-o go-template`, or `-o jsonpath`
 * Specify `--output-version`, since those output forms (other than `-o name`) output the resource using a particular API version
 * Specify `--generator` to pin to a specific behavior forever, if using generator-based commands (such as `kubectl run` or `kubectl expose`)
 * Don't rely on context, preferences, or other implicit state
 
+## Generators
+
+`kubectl run` provides following generators:
```
Comment (kargakis) [inline docs/user-guide/kubectl-conventions.md]: Why only the run generators?

**Comment by kargakis** [comment]
Is this supposed to fix https://github.com/kubernetes/kubernetes/issues/17833?

**Comment by kargakis** [comment]
As part of https://github.com/kubernetes/kubernetes/issues/17833 I had envisioned that we would document all of the existing generators and explain a bit more what is a generator (we also have some doc about the subject in https://github.com/kubernetes/kubernetes/blob/master/docs/devel/kubectl-conventions.md#generators)

**Comment by kargakis** [comment]
And call out the differences between different generators (we already have two different service generators, not sure if others exist).

**Comment by janetkuo** [comment]
This is part of https://github.com/kubernetes/kubernetes/pull/32671, which removes the lengthy description of `kubectl run --generator` and link to this doc. This PR isn't trying to include all generators

**Comment by devin-donnelly** [comment]
1.4 goes out on Monday; this can go in after 1.4 gets merged to Master (pending review changes).

File: docs/user-guide/kubectl-conventions.md (hunk 7 above)
Comment (soltysh) [inline docs/user-guide/kubectl-conventions.md]: You're free to open additional PR with more :smile:

**Comment by soltysh** [comment]
@janetkuo I've updated the doc according to your comment, ptal

File: docs/user-guide/kubectl-conventions.md [hunk 8]
```diff
@@ -29,6 +29,40 @@ In order for `kubectl run` to satisfy infrastructure as code:
 * If features are needed that are not expressible via `kubectl run` flags, switch to configuration files checked into source control.
 * Pin to a specific generator version, such as `kubectl run --generator=deployment/v1beta1`
 
+#### Generators
+
+`kubectl run` allows you to generate following resources (using `--generator` flag):
```
Comment (janetkuo) [inline docs/user-guide/kubectl-conventions.md]: nit: "the" following

File: docs/user-guide/kubectl-conventions.md [hunk 9]
```diff
@@ -29,6 +29,40 @@ In order for `kubectl run` to satisfy infrastructure as code:
 * If features are needed that are not expressible via `kubectl run` flags, switch to configuration files checked into source control.
 * Pin to a specific generator version, such as `kubectl run --generator=deployment/v1beta1`
```
Comment (janetkuo) [inline docs/user-guide/kubectl-conventions.md]: link to the "generators" section (e.g. see xxx for more)

File: docs/user-guide/kubectl-conventions.md [hunk 10]
```diff
@@ -29,6 +29,40 @@ In order for `kubectl run` to satisfy infrastructure as code:
...
+| Pod                    | `--restart=Never`     | `--restart=Never`     | `--generator=run-pod/v1`                   | `--restart=OnFailure` OR `--restart=Never` |
+| Replication Controller | `--generator=run/v1`  | `--generator=run/v1`  | `--generator=run/v1`                       | `--restart=Always`                         |
+| Deployment             | `--restart=Always`    | `--restart=Always`    | `--restart=Always`                         | N/A                                        |
+| Job                    | `--restart=OnFailure` | `--restart=OnFailure` | `--restart=OnFailure` OR `--restart=Never` | N/A                                        |
+| Scheduled Job          | `--schedule=<cron>`   | N/A                   | N/A                                        | N/A                                        |
+
+It's important to note, that these flags will default the generator only when none
+is chosen.  This also means that combining any of above flags with `--generator` will
+not change the generator at any point.
```
Comment (janetkuo) [inline docs/user-guide/kubectl-conventions.md]: "This also means that combining any of above flags with `--generator` will not change the generator at any point." --> This is a bit ambiguous, since `--generator` is also in the table. Maybe rephrase this to something like:

[...] combining `--generator` with other flags won't change the `generator` you specified. 

For example, in a 1.4 cluster, if you specify `--restart=Always`, a Deployment will be created; if you specify `--restart=Always` and `--generator=run/v1`, a Replication Controller will be created instead. This becomes handy if you want to pin to a specific behavior with the generator, even when the defaulted generator is change in the future.

File: docs/user-guide/kubectl-conventions.md (hunk 8 above)
Comment (soltysh) [inline docs/user-guide/kubectl-conventions.md]: OMG, that *always* bites me :wink: I need to write it down with big letters to remember for ever.

**Comment by soltysh** [comment]
@janetkuo updated, ptal

**Comment by janetkuo** [comment]
LGTM. GitHub says you need a rebase

**Comment by soltysh** [comment]
Rebased.

File: docs/user-guide/kubectl-conventions.md [hunk 11]
```diff
@@ -27,8 +27,46 @@ In order for `kubectl run` to satisfy infrastructure as code:
...
+`kubectl run` allows you to generate the following resources (using `--generator` flag):
+
+* Pod - use `run-pod/v1`.
+* Replication controller - use `run/v1`.
+* Deployment - use `deployment/v1beta1`.
+* Job (using `extension/v1beta1` endpoint) - use `job/v1beta1`.
+* Job - use `job/v1`.
+* ScheduledJob - use `scheduledjob/v2alpha1`.
+
+Additionally, if no generator flag was specified other flags will suggest using
```
Comment (devin-donnelly) [inline docs/user-guide/kubectl-conventions.md]: "Additionally, if you didn't specify a generator flag, other flags will suggest using a specific generator."

File: docs/user-guide/kubectl-conventions.md [hunk 12]
```diff
@@ -27,8 +27,46 @@ In order for `kubectl run` to satisfy infrastructure as code:
...
+|:----------------------:|-----------------------|-----------------------|--------------------------------------------|--------------------------------------------|
+| Pod                    | `--restart=Never`     | `--restart=Never`     | `--generator=run-pod/v1`                   | `--restart=OnFailure` OR `--restart=Never` |
+| Replication Controller | `--generator=run/v1`  | `--generator=run/v1`  | `--generator=run/v1`                       | `--restart=Always`                         |
+| Deployment             | `--restart=Always`    | `--restart=Always`    | `--restart=Always`                         | N/A                                        |
+| Job                    | `--restart=OnFailure` | `--restart=OnFailure` | `--restart=OnFailure` OR `--restart=Never` | N/A                                        |
+| Scheduled Job          | `--schedule=<cron>`   | N/A                   | N/A                                        | N/A                                        |
+
+It's important to note, that these flags will default the generator only when none
```
Comment (devin-donnelly) [inline docs/user-guide/kubectl-conventions.md]: "It's important to note," -> "Note that these flags will use a default generator only when you have not specified any flag."

File: docs/user-guide/kubectl-conventions.md [hunk 13]
```diff
@@ -27,8 +27,46 @@ In order for `kubectl run` to satisfy infrastructure as code:
...
+`--restart=Always`, a Deployment will be created; if you specify `--restart=Always`
+and `--generator=run/v1`, a Replication Controller will be created instead.
+This becomes handy if you want to pin to a specific behavior with the generator,
+even when the defaulted generator is changed in the future.
+
+Finally, the order in which flags set the generator is: schedule flag has the highest
+priority, then restart policy and finally the generator itself.
+
+If in doubt about the final resource being created, you can always use `--dry-run`
+flag which will give you the object to be submitted to cluster.
```
Comment (devin-donnelly) [inline docs/user-guide/kubectl-conventions.md]: comma after "flag"
"give you" -> "provide"
"to cluster" -> "to the cluster"

**Comment by soltysh** [comment]
@devin-donnelly comments addressed, ptal
